qiskit/circuit/quantumcircuit.py

```diff
@@ -840,98 +840,100 @@ def compose(
                 "Cannot emit a new composed circuit while a control-flow context is active."
             )
 
-        if inplace:
-            dest = self
-        else:
-            dest = self.copy()
+        dest = self if inplace else self.copy()
 
-        # If self does not have any cregs but other does then we allow
-        # that and add the registers to the output dest
+        # As a special case, allow composing some clbits onto no clbits - normally the destination
+        # has to be strictly larger. This allows composing final measurements onto unitary circuits.
         if isinstance(other, QuantumCircuit):
             if not self.clbits and other.clbits:
                 dest.add_bits(other.clbits)
                 for reg in other.cregs:
                     dest.add_register(reg)
 
-        if wrap:
-            try:
-                other = other.to_gate()
-            except QiskitError:
-                other = other.to_instruction()
+        if wrap and isinstance(other, QuantumCircuit):
+            other = (
+                other.to_gate()
+                if all(isinstance(ins.operation, Gate) for ins in other.data)
+                else other.to_instruction()
+            )
 
         if not isinstance(other, QuantumCircuit):
             if qubits is None:
-                qubits = list(range(other.num_qubits))
-
+                qubits = self.qubits[: other.num_qubits]
             if clbits is None:
-                clbits = list(range(other.num_clbits))
-
+                clbits = self.clbits[: other.num_clbits]
             if front:
-                dest.data.insert(0, CircuitInstruction(other, qubits, clbits))
+                old_data = list(dest.data)
```

Comment:
Why is rebuilding the list here required? Does the insertion at front create a bug? 🤔

Reply:
The next line does `dest.clear`, which mutates `dest.data` in place, so we'd lose access to the data if we don't clone the references.  I'd thought there was a comment, but I guess I was thinking of somewhere else in Terra that does a similar thing.  Added a comment in a8dcfb5.

Reply:
Julien mentioned on Slack that his question was about whether `data.insert(0, ...)` caused a bug, or if this was a stylistic change.  The `insert(0, x)` form is incorrect: it invalidates all existing references in the `ParameterTable`.

Reply:
Thank you!